File: example-bots/cpp-sma-crossover/query.cpp
/**
 * SMA Crossover Query Handlers (C++)
 * ===================================
 * Query endpoints for the SMA crossover bot.
 *
 * These handlers provide read-only access to bot state and computed data.
 * They can be executed via:
 *   - CLI: the0 bot query <bot_id> /status
 *   - API: POST /bot/:id/query
 *
 * Available queries:
 *   /status     - Get current SMA values and configuration
 *   /signals    - Get signal count with optional limit param
 *   /sma-values - Get current short/long SMA values
 */

#include <the0/query.h>
#include <the0/state.h>

using json = nlohmann::json;

int main() {
    /**
     * Get bot status and current SMA values.
     *
     * Returns:
     *   prev_short_sma: Previous short SMA value
     *   prev_long_sma: Previous long SMA value
     *   signal_count: Total signals generated
     */
    the0::query::handler("/status", [](const the0::query::Request& req) {
        auto state = the0::state::get("bot_state");
        auto config = the0::query::get_config();

        json result;
        if (state.has_value()) {
            result["prev_short_sma"] = state.value().value("prev_short_sma", nullptr);
            result["prev_long_sma"] = state.value().value("prev_long_sma", nullptr);
            result["signal_count"] = state.value().value("signal_count", 0);
        } else {
            result["prev_short_sma"] = nullptr;
            result["prev_long_sma"] = nullptr;
            result["signal_count"] = 0;
        }

        result["config"] = {
            {"symbol", config.value("symbol", "AAPL")},
            {"short_period", config.value("short_period", 5)},
            {"long_period", config.value("long_period", 20)}
        };

        return result;
    });

    /**
     * Get signal statistics.
     *
     * Query params:
     *   limit: Not used (for future signal history)
     *
     * Returns:
     *   signal_count: Total signals generated
     *   limit_applied: The limit parameter (for future use)
     */
    the0::query::handler("/signals", [](const the0::query::Request& req) {
        int limit = 10;
        try {
            limit = std::stoi(req.get("limit", "10"));
        } catch (...) {
            limit = 10;
        }

        auto state = the0::state::get("bot_state");
        long signal_count = state.has_value() ? state.value().value("signal_count", 0) : 0;

        return json{
            {"signal_count", signal_count},
            {"limit_applied", limit}
            // Note: Individual signals are not persisted, only the count
        };
    });

    /**
     * Get current SMA values.
     *
     * Returns:
     *   short_sma: Current short SMA value
     *   long_sma: Current long SMA value
     */
    the0::query::handler("/sma-values", [](const the0::query::Request& req) {
        auto state = the0::state::get("bot_state");

        if (state.has_value()) {
            return json{
                {"short_sma", state.value().value("prev_short_sma", nullptr)},
                {"long_sma", state.value().value("prev_long_sma", nullptr)}
            };
        }

        return json{
            {"short_sma", nullptr},
            {"long_sma", nullptr}
        };
    });

    // Run query system
    the0::query::run();
    return 0;
}

File: example-bots/csharp-sma-crossover/Query.cs
/**
 * SMA Crossover Bot - Query Handlers
 * ===================================
 * Query endpoints for the SMA crossover bot.
 *
 * These handlers provide read-only access to bot state and computed data.
 * They can be executed via:
 *   - CLI: the0 bot query <bot_id> /status
 *   - API: POST /query/:botId
 *
 * Available queries:
 *   /status  - Get bot configuration and current state
 *   /signals - Get signal statistics
 *   /sma     - Get current SMA values
 */

using The0;

class QueryProgram
{
    static void Main(string[] args)
    {
        // Register query handlers
        Query.Handler("/status", req =>
        {
            var config = Query.GetConfig();
            var botState = State.Get<BotState>("bot_state");

            return new
            {
                config = new
                {
                    symbol = config?["symbol"]?.GetValue<string>() ?? "AAPL",
                    short_period = config?["short_period"]?.GetValue<int>() ?? 5,
                    long_period = config?["long_period"]?.GetValue<int>() ?? 20,
                    update_interval_ms = config?["update_interval_ms"]?.GetValue<int>() ?? 60000
                },
                state = new
                {
                    prev_short_sma = botState?.PrevShortSma,
                    prev_long_sma = botState?.PrevLongSma,
                    signal_count = botState?.SignalCount ?? 0,
                    has_data = botState != null
                }
            };
        });

        Query.Handler("/signals", req =>
        {
            var botState = State.Get<BotState>("bot_state");

            return new
            {
                total_signals = botState?.SignalCount ?? 0,
                has_data = botState != null
            };
        });

        Query.Handler("/sma", req =>
        {
            var config = Query.GetConfig();
            var botState = State.Get<BotState>("bot_state");

            return new
            {
                symbol = config?["symbol"]?.GetValue<string>() ?? "AAPL",
                short_sma = botState?.PrevShortSma,
                long_sma = botState?.PrevLongSma,
                short_period = config?["short_period"]?.GetValue<int>() ?? 5,
                long_period = config?["long_period"]?.GetValue<int>() ?? 20
            };
        });

        // Run the query system (auto-detects mode)
        Query.Run();
    }
}

// State structure (must match the one in Program.cs)
record BotState(double? PrevShortSma, double? PrevLongSma, long SignalCount);
